lib/MusicBrainz/LWP.pm

```diff
@@ -0,0 +1,151 @@
+package MusicBrainz::LWP;
+
+use Moose;
+use namespace::autoclean;
+
+use AnyEvent;
+use AnyEvent::HTTP::LWP::UserAgent;
+use DBDefs;
+use HTTP::Status qw( HTTP_SERVICE_UNAVAILABLE );
+use LWP::UserAgent;
+use MooseX::Types::Moose qw( Int );
+use MusicBrainz::Server::Log qw( log_debug );
+use feature 'switch';
+
+no if $] >= 5.018, warnings => 'experimental::smartmatch';
+
+BEGIN { $ENV{PERL_ANYEVENT_MODEL} = 'EV'; }
+
+has 'global_timeout' => (
+    is => 'rw',
+    isa => 'Int',
+);
+
+has '_lwp_user_agent' => (
+    is => 'rw',
+    isa => 'LWP::UserAgent',
+);
+
+sub BUILD {
+    my ($self, $args) = @_;
+
+    $args->{agent} = DBDefs->LWP_USER_AGENT unless exists $args->{agent};
+
+    my $global_timeout =
+        exists $args->{global_timeout}
+            ? delete $args->{global_timeout}
+            : 0;
+
+    $self->global_timeout($global_timeout);
+    $self->_lwp_user_agent(AnyEvent::HTTP::LWP::UserAgent->new(%$args));
+}
+
+sub _timely_call_method {
+    my ($self, $lwp_method, $http_method, $arg) = @_;
+
+    my $ua = $self->_lwp_user_agent;
+    my $cv = AE::cv;
+    $cv->begin;
+
+    my $response;
+    my $global_timeout = $self->global_timeout;
+    my $got_timeout = 0;
+
+    my $uri;
+    my $request;
+    my $req_cv;
+    given ($lwp_method) {
+        when ('get') {
+            $uri = $arg;
+            $req_cv = $ua->get_async($uri);
+        }
+        when ('mirror') {
+            $uri = $arg;
+            $req_cv = $ua->mirror_async($uri);
+        }
+        when ('post') {
+            $uri = $arg;
+            $req_cv = $ua->post_async($uri);
+        }
+        when ('request') {
+            $request = $arg;
+            $req_cv = $ua->request_async($request);
+        }
+    }
+
+    $req_cv->cb(sub {
+        $response = shift->recv;
+        $cv->end;
+    });
+
+    my $timer = AnyEvent->timer(after => $global_timeout, cb => sub {
+        $req_cv->croak;
+        $got_timeout = 1;
+        $cv->croak;
+    }) unless $global_timeout == 0;
```

Comment:
For some reason this is firing immediately for me. :\ I think it was working in the original script I suggested. Maybe the PR needs to be rebased first.

Reply:
I just rebased on `master` (<https://github.com/metabrainz/musicbrainz-server/commit/36fdb6ee59b9be16d151ef799a5f33e42061a3cb>) but did not test it again yet.